masu/external/notification_handler.py

```diff
@@ -0,0 +1,120 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+"""Cost Usage Report notification handler."""
+
+from masu.external import AWS_SNS_HEADER_MSG_TYPE
+from masu.external.notifications.aws.aws_notification_handler import AWSNotificationHandler
+from masu.external.notifications.notification_interface import NotificationInterfaceFilter
+
+
+class NotificationHandlerError(Exception):
+    """Notification Handler error."""
+
+    pass
+
+
+class NotificationHandlerFilter(Exception):
+    """Notification Handler filter."""
+
+    pass
+
+
+# pylint: disable=too-few-public-methods
+# pylint: disable=too-many-arguments
+class NotificationHandler:
+    """Interface for masu to handle report download/process notifications."""
+
+    def __init__(self, header, body):
+        """
+        Set the handler based on the backend cloud provider.
+
+        Args:
+            header          ([(,)])  List of header key/value pair sets
+            body            (String) String representation of the request body JSON
+
+        Raises:
+            (NotificationHandlerError): Object fails to be initialized.
+
+        """
+        self._headers = header
+        self._body = body
+        try:
+            self._handler = self._set_handler()
+        except Exception as err:
+            raise NotificationHandlerError(str(err))
+
+        if not self._handler:
+            raise NotificationHandlerError('Unsupported cloud provider.')
+
+    def _set_handler(self):
+        """
+        Create the notification handler object.
+
+        Handler is specific to the provider's cloud service.
+
+        Args:
+            None
+
+        Returns:
+            (Object) : Some object that is a child of NotificationHandlerInterface (coming soon)
+
+        """
+        # Quick check to make sure we know what cloud the message is coming from.
+        if self._search_header_for_string(AWS_SNS_HEADER_MSG_TYPE):
+            return AWSNotificationHandler(self._headers, self._body)
+        return None
+
+    def _search_header_for_string(self, value):
+        """
+        Search the headers for a particular string.
+
+        Method is general purpose and could belong in it's own utils file but for now
+        it's a quick way to verify that the message is coming from a source that we expect.
+
+        Args:
+            None
+
+        Returns:
+            (Boolean) : True if string is found in header, else False
+
+        """
+        for header in self._headers:
+            if header[0] == value:
+                return True
+        return False
+
+    def get_notified_billing_source(self):
```

Comment:
bad code smell?  I like verbose, explicit names because you really don't know how the ```NotificationHandler``` object will be named when the method is invoked

```
notified_billing_source = handler.get_notified_billing_source()
```

Any suggestions on a better name?

Reply:
I don't mind verbose to an extent, but for me there's a point where it gets excessive. Having redundant information in there is across that line for me. (e.g. `foo_module.aws.AWSThingy.get_stuff_from_aws()` is worse, IMO, than `foo_module.aws.Thingy.stuff()`, if that helps.

To me, `.billing_source()` tells us what we need to know - that you're fetching a billing source. Though, `get_billing_source()` follows the conventions you've been using. So, that's okay, too.

Reply:
ok I'll change it